// orchextra-client-generation/customers.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var axios = require('axios');
var api = 'v1/customers';
var queryFn = require('./fn/QueryString');

module.exports = function () {
  function Customer(url, token, customer) {
    _classCallCheck(this, Customer);

    this.url = url;
    this.token = token;
    if (typeof customer !== 'undefined') {
      this.data = customer;
    }
  }

  _createClass(Customer, [{
    key: 'getCustomers',
    value: function getCustomers(query) {
      var _this = this;

      var queryString = queryFn.setQueryString(query);
      var customers = axios.get(this.url + '/' + api + '?' + queryString, {
        headers: { 'Authorization': 'Bearer ' + this.token }
      });

      return customers.then(function (result) {
        var Customers = new Array();
        result.data.forEach(function (customer) {
          Customers.push(new Customer(_this.url, _this.token, customer));
        });
        return Customers;
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }, {
    key: 'getCustomer',
    value: function getCustomer(idCustomer, query) {
      var _this2 = this;

      var queryString = queryFn.setQueryString(query);
      var customer = axios.get(this.url + '/' + api + '/' + idCustomer + '?' + queryString, {
        headers: { 'Authorization': 'Bearer ' + this.token }
      });

      return customer.then(function (result) {
        return new Customer(_this2.url, _this2.token, result.data);
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }]);

  return Customer;
}();

// module.exports = {
//   getCustomers,
//   getCustomer,
//   setUrl,
//   setAuthToken
// };
