// orchextra-client-generation/projects.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var axios = require('axios');
var api = 'v1/projects';
var queryFn = require('./fn/QueryString');

module.exports = function () {
  function Project(url, token, project) {
    _classCallCheck(this, Project);

    this.url = url;
    this.token = token;
    if (typeof project !== 'undefined') {
      this.data = project;
    }
  }

  _createClass(Project, [{
    key: 'getProjects',
    value: function getProjects(query) {
      var _this = this;

      var queryString = queryFn.setQueryString(query);
      var projects = axios.get(this.url + '/' + api + '?' + queryString, {
        headers: { 'Authorization': 'Bearer ' + this.token }
      });

      return projects.then(function (result) {
        var Projects = new Array();
        result.data.forEach(function (project) {
          Projects.push(new Project(_this.url, _this.token, project));
        });
        return Projects;
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }, {
    key: 'getProject',
    value: function getProject(idProject, query) {
      var _this2 = this;

      var queryString = queryFn.setQueryString(query);
      var project = axios.get(this.url + '/' + api + '/' + idProject + '?' + queryString, {
        headers: { 'Authorization': 'Bearer ' + this.token }
      });

      return project.then(function (result) {
        return new Project(_this2.url, _this2.token, result.data);
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }, {
    key: 'createProject',
    value: function createProject(project) {
      var _this3 = this;

      var add = axios.post(this.url + '/' + api, project, {
        headers: { 'Authorization': 'Bearer ' + this.token, 'Content-Type': 'application/json' }
      });

      return add.then(function (result) {
        return new Project(_this3.url, _this3.token, result.data);
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }, {
    key: 'deleteProject',
    value: function deleteProject(idProject) {
      if (typeof idProject === 'undefined') {
        idProject = this.data.id;
      }
      var project = axios.delete(this.url + '/' + api + '/' + idProject, {
        headers: { 'Authorization': 'Bearer ' + this.token }
      });

      return project.then(function (result) {
        return result.status;
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }, {
    key: 'updateProject',
    value: function updateProject(project, idProject) {
      var _this4 = this;

      if (typeof idProject === 'undefined') {
        idProject = this.data.id;
      }
      if (typeof project === 'undefined') {
        project = this.data;
      }
      var update = axios.patch(this.url + '/' + api + '/' + idProject, project, {
        headers: { 'Authorization': 'Bearer ' + this.token, 'Content-Type': 'application/json' }
      });

      return update.then(function (result) {
        _this4.data = result.data;
        return _this4;
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }, {
    key: 'replaceProject',
    value: function replaceProject(project, idProject) {
      var _this5 = this;

      if (typeof idProject === 'undefined') {
        idProject = this.data.id;
      }
      var update = axios.put(this.url + '/' + api + '/' + idProject, project, {
        headers: { 'Authorization': 'Bearer ' + this.token, 'Content-Type': 'application/json' }
      });

      return update.then(function (result) {
        _this5.data = result.data;
        return _this5;
      }).catch(function (error) {
        return {
          statusCode: error.response.status,
          errors: error.response.data
        };
      });
    }
  }]);

  return Project;
}();

// module.exports = {
//   getProjects,
//   getProject,
//   createProject,
//   deleteProject,
//   updateProject,
//   replaceProject,
//   setUrl,
//   setAuthToken
// };
